spline/bezier.go

```diff
@@ -0,0 +1,172 @@
+// TODO:
+// 1: type Bezier (points and control points) (done)
+// 2: open to factors 5th-order Bezier (not needed because 5 is done)
+// 3: derivatives (done)
+// 4: get/set optimization (done)
+
+// 5: general n-th order open to factors (done)
+// 6: smoothing with other Bezier (not done yet)
+
+package spline
+
+import (
+	"math"
+)
+
+//this includes edge points and control points
+//when smoothing with other bezier we might need to seperate edge from control points
+type BezierPoint struct {
+	X float64
+	Y float64
+}
+
+type Bezier struct {
+	points []BezierPoint
+
+	xFactors   []float64
+	dXFactors  []float64
+	ddXFactors []float64
+
+	yFactors   []float64
+	dYFactors  []float64
+	ddYFactors []float64
+
+	optimizationParamsLength uint
+}
+
+func newBezierCurve(points []BezierPoint) Bezier {
+	bezier := Bezier{points: points, optimizationParamsLength: uint((len(points) - 2) * 2)}
+	xFactors := bezier.GetXFactors()
+	yFactors := bezier.GetYFactors()
+
+	bezier.SetXFactors(xFactors)
+	bezier.SetYFactors(yFactors)
+
+	return bezier
+}
+
+func (bezier Bezier) SetXFactors(xFactors []float64) {
+	bezier.xFactors = xFactors
+	bezier.dXFactors = getDerivativeFactors(xFactors)
+	bezier.ddXFactors = getDerivativeFactors(bezier.dXFactors)
+}
+
+func (bezier Bezier) SetYFactors(yFactors []float64) {
+	bezier.yFactors = yFactors
+	bezier.dYFactors = getDerivativeFactors(yFactors)
+	bezier.ddYFactors = getDerivativeFactors(bezier.dYFactors)
+}
+
+//GetXFactors and GetYFactors are the exact same but the only difference is in lines 70 and 86
```

Comment:
Think of a way to write these function in a more elegant way - using a utility function to reduce the amount of duplicated code